sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3805,3 +3799,233 @@ object ArrayUnion {
     new GenericArrayData(arrayBuffer)
   }
 }
+
+/**
+ * Returns an array of the elements in the intersect of x and y, without duplicates
+ */
+@ExpressionDescription(
+  usage = """
+  _FUNC_(array1, array2) - Returns an array of the elements in array1 but not in array2,
+    without duplicates.
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3), array(1, 3, 5));
+       array(2)
+  """,
+  since = "2.4.0")
+case class ArrayExcept(left: Expression, right: Expression) extends ArraySetLike {
```

Comment:
does this need to extend `ComplexTypeMergingExpression`?

Reply:
The check at `dataType` in `ComplexTypeMergingExpression` should be useful.
Based on [the discussion](https://github.com/apache/spark/pull/21103#discussion_r203746577), the result of `dataType should be only `left.dataType`. Will we use only the checks?

Reply:
we can overwrite `dataType` while still extending `ComplexTypeMergingExpression` to use the checks.

Reply:
Maybe I'm missing something, but why we need to apply these checks if there won't be any ```null``` flag merging performed?

If ```left.dataType``` and ```right.dataType``` are different, will be casted according to the ```ImplicitTypeCasts``` coercion rule. If they differ only in ```null``` flags, ```left.dataType``` could be directly returned since there won't be any array elements from ```right``` present in the result.